### PR TITLE
fix: do not force unwrap accessError SQPIT-798

### PR DESCRIPTION
**Pull request by billypchan**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQPIT-798" title="SQPIT-798" target="_blank">TaskSQPIT-798</a>  Azure: data model test crashed but report as passed
  </td></table>
  <br />
 

# What's new in this PR?

### Issues

Test crashes when running `testThatItStoresAndClearsDatabaseKeyOnAllContexts`

### Causes (Optional)

When on iOS 15 simulator, `accessError` is nil 

### Solutions

Do not force unwrap `accessError`, throw  case `failedToGenerateAccountKey` with error is nil
In `testThatItStoresAndClearsDatabaseKeyOnAllContexts`, use `XCTUnwrap` to fail earlier when `EncryptionKeys.createKeys` throws

These tests failed as expected with iOS 15 simulator:

```
Failing tests:
	WireDataModelTests:
		CoreDataStackTests_EncryptionAtRest.testThatItStoresAndClearsDatabaseKeyOnAllContexts()
		CoreDataStackTests_EncryptionAtRest.testThatItStoresAndClearsDatabaseKeyOnAllContexts()
		EncryptionKeysTests.testThatAsymmetricKeysWorksWithExpectedAlgorithm()
		EncryptionKeysTests.testThatEncryptionKeysAreSuccessfullyCreated()
		EncryptionKeysTests.testThatEncryptionKeysAreSuccessfullyDeleted()
		EncryptionKeysTests.testThatEncryptionKeysAreSuccessfullyFetched()
		EncryptionKeysTests.testThatPublicAccountKeyIsReturnedIfItExists()

```

Added XCTExpectFailure for those failing tests if the simulator is iOS 15
Added GitHub action for testing with iOS 14 simulator to confirmed the issue only occurs on iOS 15 simulator.

### CODE REVIEW

**Review by sb88k — APPROVED** [review]

**Review by sb88k — APPROVED** [review]

**Review by typfel — CHANGES_REQUESTED** [review]
Let's add reference to https://wearezeta.atlassian.net/browse/SQCORE-1188 which contains the context on why these tests fail on iOS 15.

File: Tests/Source/ManagedObjectContext/CoreDataStackTests+EncryptionAtRest.swift [hunk 1]
```diff
@@ -22,11 +22,16 @@ import XCTest
...
     // @SF.Storage @TSFI.UserInterface @S0.1 @S0.2
-    func testThatItStoresAndClearsDatabaseKeyOnAllContexts() {
+    func testThatItStoresAndClearsDatabaseKeyOnAllContexts() throws {
         // Given
         let sut = createStorageStackAndWaitForCompletion()
         let account = Account(userName: "", userIdentifier: UUID())
-        let encryptionKeys = try! EncryptionKeys.createKeys(for: account)
+#if targetEnvironment(simulator)
+        if #available(iOS 15, *) {
+            XCTExpectFailure("Expect to fail on iOS 15 simulator.")
```
Comment (billypchan) [inline Tests/Source/ManagedObjectContext/CoreDataStackTests+EncryptionAtRest.swift:31]:
```suggestion
            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")
```

File: Tests/Source/Utils/EncryptionKeysTests.swift [hunk 2]
```diff
@@ -44,6 +44,11 @@ class EncryptionKeysTests: XCTestCase {
 
     func testThatPublicAccountKeyIsReturnedIfItExists() throws {
         // given
+#if targetEnvironment(simulator)
+        if #available(iOS 15, *) {
+            XCTExpectFailure("Expect to fail on iOS 15 simulator.")
```
Comment (billypchan) [inline Tests/Source/Utils/EncryptionKeysTests.swift:49]:
```suggestion
            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")```

File: Tests/Source/Utils/EncryptionKeysTests.swift [hunk 3]
```diff
@@ -56,6 +61,11 @@ class EncryptionKeysTests: XCTestCase {
     // @SF.Storage @TSFI.UserInterface @S0.1 @S0.2
     func testThatEncryptionKeysAreSuccessfullyCreated() throws {
         // when
+#if targetEnvironment(simulator)
+        if #available(iOS 15, *) {
+            XCTExpectFailure("Expect to fail on iOS 15 simulator.")
```
Comment (billypchan) [inline Tests/Source/Utils/EncryptionKeysTests.swift:66]:
```suggestion
            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")```

File: Tests/Source/Utils/EncryptionKeysTests.swift [hunk 4]
```diff
@@ -89,6 +109,11 @@ class EncryptionKeysTests: XCTestCase {
     func testThatAsymmetricKeysWorksWithExpectedAlgorithm() throws {
         // given
         let data = "Hello world".data(using: .utf8)!
+#if targetEnvironment(simulator)
+        if #available(iOS 15, *) {
+            XCTExpectFailure("Expect to fail on iOS 15 simulator.")
```
Comment (billypchan) [inline Tests/Source/Utils/EncryptionKeysTests.swift:114]:
```suggestion
            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")```

File: Tests/Source/Utils/EncryptionKeysTests.swift [hunk 5]
```diff
@@ -76,6 +91,11 @@ class EncryptionKeysTests: XCTestCase {
     // @SF.Storage @TSFI.UserInterface @S0.1 @S0.2
     func testThatEncryptionKeysAreSuccessfullyDeleted() throws {
         // given
+#if targetEnvironment(simulator)
+        if #available(iOS 15, *) {
+            XCTExpectFailure("Expect to fail on iOS 15 simulator.")
```
Comment (billypchan) [inline Tests/Source/Utils/EncryptionKeysTests.swift:96]:
```suggestion
            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")```

File: Tests/Source/Utils/EncryptionKeysTests.swift [hunk 6]
```diff
@@ -64,6 +74,11 @@ class EncryptionKeysTests: XCTestCase {
 
     func testThatEncryptionKeysAreSuccessfullyFetched() throws {
         // given
+#if targetEnvironment(simulator)
+        if #available(iOS 15, *) {
+            XCTExpectFailure("Expect to fail on iOS 15 simulator.")
```
Comment (billypchan) [inline Tests/Source/Utils/EncryptionKeysTests.swift:79]:
```suggestion
            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")```

File: Source/Utilis/EncryptionKeys.swift [hunk 7]
```diff
@@ -271,7 +271,7 @@ public struct EncryptionKeys {
         
         var error: Unmanaged<CFError>?
         guard let privateKey = SecKeyCreateRandomKey(attributes as CFDictionary, &error) else {
-            let error = accessError!.takeRetainedValue() as Error
+            let error = accessError?.takeRetainedValue()
```
Comment (billypchan) [inline Source/Utilis/EncryptionKeys.swift:274]:
```suggestion
            //Notice: accessError is nil when test with iOS 15 simulator. ref:https://wearezeta.atlassian.net/browse/SQCORE-1188
            let error = accessError?.takeRetainedValue()
```

File: Tests/Source/Utils/EncryptionKeysTests.swift [hunk 8]
```diff
@@ -44,6 +44,11 @@ class EncryptionKeysTests: XCTestCase {
 
     func testThatPublicAccountKeyIsReturnedIfItExists() throws {
         // given
+#if targetEnvironment(simulator)
+        if #available(iOS 15, *) {
+            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")```
```
Comment (billypchan) [inline Tests/Source/Utils/EncryptionKeysTests.swift:49]:
```suggestion
            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")
```

File: Tests/Source/Utils/EncryptionKeysTests.swift [hunk 9]
```diff
@@ -56,6 +61,11 @@ class EncryptionKeysTests: XCTestCase {
     // @SF.Storage @TSFI.UserInterface @S0.1 @S0.2
     func testThatEncryptionKeysAreSuccessfullyCreated() throws {
         // when
+#if targetEnvironment(simulator)
+        if #available(iOS 15, *) {
+            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")```
```
Comment (billypchan) [inline Tests/Source/Utils/EncryptionKeysTests.swift:66]:
```suggestion
            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")
```

File: Tests/Source/Utils/EncryptionKeysTests.swift [hunk 10]
```diff
@@ -64,6 +74,11 @@ class EncryptionKeysTests: XCTestCase {
 
     func testThatEncryptionKeysAreSuccessfullyFetched() throws {
         // given
+#if targetEnvironment(simulator)
+        if #available(iOS 15, *) {
+            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")```
```
Comment (billypchan) [inline Tests/Source/Utils/EncryptionKeysTests.swift:79]:
```suggestion
            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")
```

File: Tests/Source/Utils/EncryptionKeysTests.swift [hunk 11]
```diff
@@ -76,6 +91,11 @@ class EncryptionKeysTests: XCTestCase {
     // @SF.Storage @TSFI.UserInterface @S0.1 @S0.2
     func testThatEncryptionKeysAreSuccessfullyDeleted() throws {
         // given
+#if targetEnvironment(simulator)
+        if #available(iOS 15, *) {
+            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")```
```
Comment (billypchan) [inline Tests/Source/Utils/EncryptionKeysTests.swift:96]:
```suggestion
            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")
```

File: Tests/Source/Utils/EncryptionKeysTests.swift [hunk 12]
```diff
@@ -89,6 +109,11 @@ class EncryptionKeysTests: XCTestCase {
     func testThatAsymmetricKeysWorksWithExpectedAlgorithm() throws {
         // given
         let data = "Hello world".data(using: .utf8)!
+#if targetEnvironment(simulator)
+        if #available(iOS 15, *) {
+            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")```
```
Comment (billypchan) [inline Tests/Source/Utils/EncryptionKeysTests.swift:114]:
```suggestion
            XCTExpectFailure("Expect to fail on iOS 15 simulator. ref: https://wearezeta.atlassian.net/browse/SQCORE-1188")
```

**Review by johnxnguyen — APPROVED** [review]
It'd be nice to fix the indentation of the `#if` blocks.

**Review by typfel — APPROVED** [review]

**Comment by billypchan** [comment]
release major

**Comment by zenkins** [comment]
Hello, @billypchan! Merging PR for command **release major** for PR #**1308**. Will create a **major** release after merged.
Next version is 280.0.0.

**Pull request merged**